prae/src/core.rs

```diff
@@ -1,35 +1,105 @@
+pub struct ConstructionError<G: Guard> {
+    /// Original error in case of calling [`define!`](prae_macro::define) with `validate` or just a
+    /// string in case of `ensure`.
+    pub inner: G::Error,
+    /// The value that caused the error.
+    pub value: G::Target,
 }
 
-impl ValidationError {
-    /// Create a new error with the input value that failed.
-    pub fn new(type_name: &'static str, value: String) -> Self {
-        ValidationError { type_name, value }
+// FIXME: the compiler thinks that `G::Error` can be `ConstructionError<G>`,
+// which conflicts with default implementation From<T> for T.
+// I have no idea how to fix it!
+// impl<G: Guard> From<ConstructionError<G>> for G::Error {
+//     fn from(err: ConstructionError<G>) -> Self {
+//         err.inner
+//     }
+// }
+
+impl<G: Guard> fmt::Display for ConstructionError<G>
+where
+    G::Error: fmt::Debug + fmt::Display,
+    G::Target: fmt::Debug,
+    G: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "failed to create {} from value {:?}: {}",
+            guard_alias_name::<G>(),
+            self.value,
+            self.inner,
+        )
     }
 }
 
-impl Error for ValidationError {}
+impl<G: Guard> error::Error for ConstructionError<G>
+where
+    G::Error: fmt::Debug + fmt::Display,
+    G::Target: fmt::Debug,
+    G: fmt::Debug,
+{
+}
 
-impl fmt::Display for ValidationError {
+/// An error occured during [mutation](Guarded::try_mutate).
+#[derive(Debug)]
+pub struct MutationError<G: Guard> {
+    /// Original error in case of calling [`define!`](prae_macro::define) with `validate` or just a
+    /// string in case of `ensure`.
+    pub inner: G::Error,
+    /// The value before mutation.
+    pub old_value: G::Target,
+    /// The value that caused the error.
+    pub new_value: G::Target,
+}
+
+// FIXME: the compiler thinks that `G::Error` can be `MutationError<G>`,
+// which conflicts with default implementation From<T> for T.
+// I have no idea how to fix it!
+// impl<G: Guard> From<MutationError<G>> for G::Error {
+//     fn from(err: MutationError<G>) -> Self {
+//         err.inner
+//     }
+// }
+
+impl<G: Guard> fmt::Display for MutationError<G>
+where
+    G::Error: fmt::Debug + fmt::Display,
+    G::Target: fmt::Debug,
+    G: fmt::Debug,
+{
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         write!(
             f,
-            "failed to create {} from value {}: provided value is invalid",
-            self.type_name, self.value,
+            "failed to mutate {} from value {:?} to {:?}: {}",
+            guard_alias_name::<G>(),
+            self.old_value,
+            self.new_value,
+            self.inner,
         )
     }
 }
 
+impl<G: Guard> error::Error for MutationError<G>
+where
+    G::Error: fmt::Debug + fmt::Display,
+    G::Target: fmt::Debug,
+    G: fmt::Debug,
+{
+}
+
+/// Dirty hack to get "Name" instead of "module::NameGuard".
```

Comment:
I was thinking to that we could implement a `fn` on `Guard` that returns the alias name, since an instance of `Guard` is already created in the proc macro.

```rust
let guard_ident = format_ident!("{}Guard", ident);
let output = quote! {
   #[derive(Debug)]
    #vis struct #guard_ident;
    impl prae::Guard for #guard_ident {
        type Target = #ty;
        type Error = #err_ty;
        #adjust_fn
        #validate_fn

        fn alias_name() -> &'static str {
            #ident
        }
    }
    #vis type #ident = prae::Guarded<#guard_ident>;
};
```

Reply:
Wow, that's perfect!

Reply:
I'm glad you like it :laughing: